jupyter-org-extensions.el

```diff
@@ -0,0 +1,404 @@
+;;; jupyter-org-extensions.el --- Jupyter Org Extensions -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2018 Nathaniel Nicandro
+
+;; Author: Carlos Garcia C. <carlos@binarycharly.com>
+;; Created: 01 March 2019
+;; Version: 0.7.1
+
+;; This program is free software; you can redistribute it and/or
+;; modify it under the terms of the GNU General Public License as
+;; published by the Free Software Foundation; either version 2, or (at
+;; your option) any later version.
+
+;; This program is distributed in the hope that it will be useful, but
+;; WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+;; General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with GNU Emacs; see the file COPYING.  If not, write to the
+;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+;; Boston, MA 02111-1307, USA.
+
+;;; Commentary:
+
+;; Functions that extend the functionality of Org mode to interact with
+;; jupyter source blocks.
+
+;;; Code:
+
+(require 'org)
+(require 'jupyter-org-client)
+
+(declare-function org-babel-jupyter-initiate-session "ob-jupyter" (&optional session params))
+
+;;;###autoload
+(defun jupyter-org-insert-src-block (&optional below)
+  "Insert a src block above the current point.
+With prefix arg BELOW, insert it below the current point.
+
+If point is in a block, copy the header to the new block"
+  (interactive "P")
+  (if (org-in-src-block-p)
+      (let* ((src (org-element-context))
+             (start (org-element-property :begin src))
+             (end (org-element-property :end src))
+             (lang (org-element-property :language src))
+             (switches (org-element-property :switches src))
+             (parameters (org-element-property :parameters src))
+             location)
+        (if below
+            (progn
+              (goto-char start)
+              (setq location (org-babel-where-is-src-block-result))
+              (if (not location)
+                  (goto-char end)
+                (goto-char location)
+                (goto-char (org-element-property :end (org-element-context))))
+              (insert
+               (org-element-interpret-data
+                (org-element-put-property
+                 (jupyter-org-src-block lang parameters "\n" switches)
+                 :post-blank 1)))
+              (forward-line -3))
+          ;; after current block
+          (goto-char (org-element-property :begin src))
+          (insert
+           (org-element-interpret-data
+            (org-element-put-property
+             (jupyter-org-src-block lang parameters "\n" switches)
+             :post-blank 1)))
+          (forward-line -3)))
+
+    ;; not in a src block, insert a new block, query for jupyter kernel
+    (beginning-of-line)
+    (let ((kernel
+           (completing-read "Kernel: "
```

Comment:
There is also the function `jupyter-completing-read-kernelspec` which does the same thing.

Reply:
Changed to use `jupyter-completing-read-kernelspec`.